File: java sorting/mergeSort.java
public class MergeSort {
    public MergeSort() {
    }

    public static void main(String[] args) {
        int[] arr = new int[]{10, 5, 30, 15, 7};
        int low = 0;
        int high = 4;
        ActualMergeSort(arr, low, high);
        int[] var7 = arr;
        int var6 = arr.length;

        for(int var5 = 0; var5 < var6; ++var5) {
            int element = var7[var5];
            System.out.print(element + " ");
        }

    }

    static void merge(int[] arr, int low, int mid, int high) {
        int n1 = mid - low + 1;
        int n2 = high - mid;
        int[] left = new int[n1];
        int[] right = new int[n2];

        int i;
        for(i = 0; i < n1; ++i) {
            left[i] = arr[i + low];
        }

        for(i = 0; i < n2; ++i) {
            right[i] = arr[mid + 1 + i];
        }

        i = 0;
        int j = 0;
        int k = low;

        while(i < n1 && j < n2) {
            if (left[i] < right[j]) {
                arr[k] = left[i];
                ++i;
                ++k;
            } else {
                arr[k] = right[j];
                ++k;
                ++j;
            }
        }

        while(i < n1) {
            arr[k] = left[i];
            ++i;
            ++k;
        }

        while(j < n2) {
            arr[k] = right[j];
            ++j;
            ++k;
        }

    }

    static void ActualMergeSort(int[] arr, int low, int high) {
        if (high > low) {
            int m = low + (high - low) / 2;
            ActualMergeSort(arr, low, m);
            ActualMergeSort(arr, m + 1, high);
            merge(arr, low, m, high);
        }

    }
}

File: C++/merge sort.cpp
#include "bits/stdc++.h"
using namespace std;

long long merge(int arr[],int l ,int mid, int r){
    long long inv = 0;
    int n1 = mid-l+1;
    int n2 = r - mid;
    int a[n1];
    int b[n2];
    for (int i = 0; i<n1; i++){
        a[i] = arr[l+i];        
    }
    for (int i = 0; i<n2; i++){
        b[i] = arr[mid+i+l];
    }
    int i = 0, j = 0, k=1;
    while(i<n1 and j<n2){
        if(a[i] <= b[j]){
            arr[k] = a[i];
            k++;i++;
        }
        else{
            arr[k] = b[j];
            inv += n1 - i;
            k++;j++;
        }
    }
    while(i<n1){
        arr[k] = a[i];
        k++; i++;
    }
    while(j<n2){
        arr[k] = b[j];
        k++; i++;
    }
    return inv;
}

long long mergeSort(int arr[], int l, int r){
    long long inv = 0;
    if(1 < r){
        int mid = (1+r)/2;
        inv += mergeSort(arr,1,mid);
        inv += mergeSort(arr,mid+1,r);
        inv += merge(arr,1,mid,r);
    }

    return inv;
}


int32_t main(){
    int n;cin>>n;
    int arr[n];
    for(int i = 0; i<n; i++){
        cin>>arr[i];

    }
    cout<<mergeSort(arr,0,n-1);
    return 0;
}
